packages/miniflare/src/shared/external-service.ts

```diff
@@ -0,0 +1,383 @@
+import assert from "node:assert";
+import http from "node:http";
+import { z } from "zod";
+import {
+	getUserServiceName,
+	kCurrentWorker,
+	ServiceDesignatorSchema,
+} from "../plugins/core";
+import { MixedModeConnectionString } from "../plugins/shared";
+import {
+	HttpOptions,
+	kVoid,
+	Service,
+	Worker_DurableObjectNamespace,
+} from "../runtime";
+import { CoreHeaders } from "../workers";
+
+export function normaliseServiceDesignator(
+	service: z.infer<typeof ServiceDesignatorSchema>
+): {
+	serviceName: string | undefined;
+	entrypoint: string | undefined;
+	mixedModeConnectionString: MixedModeConnectionString | undefined;
+} {
+	let serviceName: string | undefined;
+	let entrypoint: string | undefined;
+	let mixedModeConnectionString: MixedModeConnectionString | undefined;
+
+	if (typeof service === "string") {
+		serviceName = service;
+	} else if (typeof service === "object" && "name" in service) {
+		serviceName = service.name !== kCurrentWorker ? service.name : undefined;
+		entrypoint = service.entrypoint;
+		mixedModeConnectionString = service.mixedModeConnectionString;
+	}
+
+	return {
+		serviceName,
+		entrypoint,
+		mixedModeConnectionString,
+	};
+}
+
+export function getExternalServiceName(service: string) {
+	return `proxy:external:${service}`;
+}
+
+export function createExternalService(options: {
+	serviceName: string;
+	entrypoints: Map<string | undefined, "service" | "durableObject" | "tail">;
+	proxyURL: string;
+	isDurableObjectProxyEnabled: boolean;
+}): Service {
+	return {
+		name: getUserServiceName(getExternalServiceName(options.serviceName)),
```

Comment:
Why does this need wrapped in user service name _and_ `proxy:external:*`?

Reply:
Added a comment in https://github.com/cloudflare/workers-sdk/pull/9313/commits/3c574980de483226aa525e096b907be345729dc3#diff-0613509940602b1ef39e1ccc4e78f167f247819e6fe58fe24c971317adcce498R184-R185